google/cloud/aiplatform/datasets.py

```diff
@@ -15,6 +15,358 @@
 # limitations under the License.
 #
 
+from typing import Optional, Sequence, Union, List, Dict, Tuple
 
-class Dataset:
-    pass
+from google.api_core import operation
+from google.auth import credentials as auth_credentials
+
+from google.cloud.aiplatform import base
+from google.cloud.aiplatform import utils
+from google.cloud.aiplatform import initializer
+from google.cloud.aiplatform.utils import validate_id
+from google.cloud.aiplatform.utils import validate_string_list
+from google.cloud.aiplatform.utils import extract_fields_from_resource_name
+
+from google.cloud.aiplatform_v1beta1 import GcsSource
+from google.cloud.aiplatform_v1beta1 import GcsDestination
+from google.cloud.aiplatform_v1beta1 import ExportDataConfig
+from google.cloud.aiplatform_v1beta1 import ImportDataConfig
+from google.cloud.aiplatform_v1beta1 import ImportDataRequest
+from google.cloud.aiplatform_v1beta1 import Dataset as GapicDataset
+from google.cloud.aiplatform_v1beta1 import DatasetServiceClient
+
+
+# TODO(b/171275584): Add async support to Dataset class
+class Dataset(base.AiPlatformResourceNoun):
+    """Managed dataset resource for AI Platform"""
+
+    client_class = DatasetServiceClient
+    _is_client_prediction_client = False
+
+    def __init__(
+        self,
+        dataset_name: str,
+        project: Optional[str] = None,
+        location: Optional[str] = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ):
+        """Retrieves an existing managed dataset given a dataset name or ID.
+
+        Args:
+            dataset_name (str):
+                Required. A fully-qualified dataset resource name or dataset ID.
+                Example: "projects/123/locations/us-central1/datasets/456" or
+                "456" when project and location are initialized or passed.
+            project (str):
+                Optional project to retrieve dataset from. If not set, project
+                set in aiplatform.init will be used.
+            location (str):
+                Optional location to retrieve dataset from. If not set, location
+                set in aiplatform.init will be used.
+            credentials: Optional[auth_credentials.Credentials]=None,
+                Custom credentials to use to upload this model. Overrides
+                credentials set in aiplatform.init.
+        """
+
+        # Fully qualified dataset name, i.e. "projects/.../locations/.../datasets/12345"
+        valid_name = extract_fields_from_resource_name(
+            resource_name=dataset_name, resource_noun="datasets"
+        )
+
+        # Partial dataset name (i.e. "12345") with known project and location
+        if (
```

Comment:
will something similar to this block be needed for every resource?  if so let's refactor it into utils or the base class later.